### PR TITLE
Small refactoring

### PR DESCRIPTION
**Pull request by darkdef**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ✔️

### CODE REVIEW

**Comment by samdark** [comment]
@darkdef please update from master.

File: src/Category.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<?php
```
Comment (romkatsu) [inline src/Category.php:1]: set strict type please

File: src/Category.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace Yiisoft\Translator;
+
+final class Category
+{
+    /**
+     * @var MessageReaderInterface
+     */
+    private $reader;
```
Comment (romkatsu) [inline src/Category.php:10]: set type please

File: src/Translator.php [hunk 3]
```diff
@@ -4,105 +4,68 @@
...
+        $message = $this->categories[$category]->getReader()->getMessage($id, $category, $locale, $parameters);
         if ($message === null) {
-            $missingTranslation = new MissingTranslationEvent($category, $localeString, $id);
+            $missingTranslation = new MissingTranslationEvent($category, $locale, $id);
             $this->eventDispatcher->dispatch($missingTranslation);
 
-            $locale = new Locale($localeString);
-            $fallback = $locale->fallbackLocale();
+            $localeObject = new Locale($locale);
+            $fallback = $localeObject->fallbackLocale();
```
Comment (samdark) [inline src/Translator.php:57]: Fallback should be done multiple times until translation is either found or not found anywhere. For example, `hy-Latn-IT-arevela` falls back to `hy-Latn-IT` then to `hy-Latn` then to `hy`.

File: src/Translator.php [hunk 4]
```diff
@@ -4,105 +4,68 @@
...
-            $localeString = $this->getLocale();
+        $locale = $locale ?? $this->defaultLocale;
+        if (empty($locale)) {
+            return $id;
         }
 
-        if ($category === null) {
-            $category = $this->getDefaultCategory();
+        $category = $category ?? $this->defaultCategory;
+        if (empty($category) or empty($this->categories[$category])) {
```
Comment (samdark) [inline src/Translator.php:47]:
```suggestion
        if (empty($category) || empty($this->categories[$category])) {
```

File: src/TranslatorInterface.php [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+interface TranslatorInterface
+{
+    /**
+     * Translates a message to the specified language.
+     *
+     * @param string $id the id of the message to be translated. It can be either artificial ID or the source message.
+     * @param array  $parameters An array of parameters for the message
+     * @param string $category the message category
+     * @param string $locale the target locale
```
Comment (samdark) [inline src/TranslatorInterface.php:15]:
```suggestion
     * @param string|null $category the message category
     * @param string|null $locale the target locale
```

**Review by samdark — CHANGES_REQUESTED** [review]

File: src/Translator.php (hunk 3 above)
Comment (samdark) [reply src/Translator.php:57]: A test should be implemented for this case.

File: src/Translator.php (hunk 4 above)
Comment (darkdef) [reply src/Translator.php]: fixed

File: src/TranslatorInterface.php (hunk 5 above)
Comment (darkdef) [reply src/TranslatorInterface.php]: fixed

File: src/Translator.php (hunk 3 above)
Comment (darkdef) [reply src/Translator.php:57]: Tests added

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thank you! 👍

**Pull request merged**